keps/sig-cluster-lifecycle/draft-20180316-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,243 @@
+status: accepted
+authors:
+  - "@fabriziopandini"
+owning-sig: sig-cluster-lifecycle
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor:
+  - "@fabriziopandini"
+creation-date: 2018-03-16
+last-updated: 2018-03-16
+see-also:
+
+```
+
+## Table of Contents
+
+  * [kubeadm phases to beta (GA)](#kubeadm-phases-to-beta-ga)
+    * [Metadata](#metadata)
+    * [Table of Contents](#table-of-contents)
+    * [Summary](#summary)
+    * [Motivation](#motivation)
+        * [Goals](#goals)
+        * [Non-goals](#non-goals)
+        * [Challenges and Open Questions](#challenges-and-open-questions)
+    * [Proposal](#proposal)
+        * [User Stories](#user-stories)
+          * [Invoke a specific task/phase of the kubeadm init workflow](#invoke-a-specific-taskphase-of-the-kubeadm-init-workflow)
+        * [Implementation Details](#implementation-details)
+          * [Expected benefits](#expected-benefits)
+          * [Phased Command Builder](#phased-command-builder)
+          * [cmdUtil factory](#cmdutil-factory)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Implementation History](#implementation-history)
+    * [Drawbacks](#drawbacks)
+    * [Alternatives](#alternatives)
+
+## Summary
+
+We are defining the roadmap for graduating `kubeadm alpha phase` commands to
+beta (and GA), addressing concerns/lessons learned so far about the additional
+effort for maintenance of this feature.
+
+## Motivation
+
+The `kubeadm phase` command was introduced in v1.8 cycle under `kubeadm alpha phase`
+command with the goal to provide to users an interface for invoking individually any
+task/phase of the `kubeadm init` workflow.
+
+During this period,  `kubeadm phase`  proved to be a valuable and composable
+API/toolbox that can be used by any IT automation tool or by advanced user for
+creating custom clusters.
+
+On the other side, it was necessary a continuous effort for maintaining this
+component up to date with the changes of  `kubeadm init`, especially with regards
+to the proliferation of flags and the duplication of code and, as a consequence, with
+the risk of potential inconsistencies between init and phases implementation logic.
+
+This KEP aims to define the way forward for graduating `kubeadm alpha phase`
+commands to beta (and GA) and, at the same time, to enable a significant reduction
+of the effort for maintaining this feature.
+
+### Goals
+
+* To define the approach for graduating to beta (GA) the `kubeadm phase` user
+  interface.
+* To define a new approach for implementing the `kubeadm phase` user interface
+  that significantly reduces the effort for maintaining this feature up to date
+  with the continuous changes of the init workflow.
+* To provide a centralized point for controlling the proliferation of flags in
+  kubeadm.
+* To avoid inconsistencies between init and phases (both at UX and implementation
+  logic level)
+
+### Non-goals
+
+* This proposal doesn't include any changes/improvement to the init workflow
```

Comment:
how is modifying `kubeadm init` to run through the phases not a change to the "init workflow internal logic?" is something else meant by this?

Reply:
Even if this KEP imply a refactoring of kubeadm init for allowing execution of the entire workflow or single phases, when you will execute the entire workflow you will get the same result as today e.g. the same certs will be created, the same manifest file will be created etc.